// RUN_PIPELINE_TILL: BACKEND
// FIR_IDENTICAL
// IGNORE_FIR_DIAGNOSTICS
// DIAGNOSTICS: -NOTHING_TO_INLINE
// NO_CHECK_LAMBDA_INLINING

private fun funOK() = "OK"

internal inline fun internalInlineFun(ok: () -> String = <!IR_PRIVATE_CALLABLE_REFERENCED_BY_NON_PRIVATE_INLINE_FUNCTION_ERROR!>::<!CALLABLE_REFERENCE_TO_LESS_VISIBLE_DECLARATION_IN_INLINE_ERROR!>funOK<!><!>) = ok()
