// FIR_IDENTICAL
// DIAGNOSTICS: -CONTEXT_RECEIVERS_DEPRECATED
// WITH_STDLIB
// RENDER_ALL_DIAGNOSTICS_FULL_TEXT
// IGNORE_FIR_DIAGNOSTICS
// IGNORE_ERRORS
// LANGUAGE: +ContextParameters

<!CONFLICTING_JVM_DECLARATIONS!>object O {
    <!CONTEXT_PARAMETERS_WITH_BACKING_FIELD!>context<!>(_: O)
    @JvmField
    <!CONFLICTING_JVM_DECLARATIONS!>val INSTANCE: O? = null<!>
}<!>
